mariner/file_formats/cbddlp.py

```diff
@@ -0,0 +1,192 @@
+                x = 0
+                array.append([])
+
+    array.pop()
+
+    return png.from_array(array, "RGB;5")
+
+
+@dataclass(frozen=True)
+class CBDDLPFile:
+    filename: str
+    bed_size_mm: Tuple[float, float, float]
+    height_mm: float
+    layer_height_mm: float
+    layer_count: int
+    resolution: Tuple[int, int]
+    print_time_secs: int
+    end_byte_offset_by_layer: Sequence[int]
+    slicer_version: str
+    printer_name: str
+
+    @classmethod
+    def read(self, path: pathlib.Path) -> "CBDDLPFile":
+        with open(str(path), "rb") as file:
+            CBDDLP_header = CBDDLPHeader.unpack(file.read(CBDDLPHeader.get_size()))
+
+            file.seek(CBDDLP_header.slicer_offset)
+            CBDDLP_slicer = CBDDLPSlicer.unpack(file.read(CBDDLPSlicer.get_size()))
+
+            file.seek(CBDDLP_slicer.machine_offset)
+            printer_name = file.read(CBDDLP_slicer.machine_size).decode()
+
+            end_byte_offset_by_layer = []
+            for layer in range(0, CBDDLP_header.layer_count):
+                file.seek(CBDDLP_header.layer_defs_offset + layer * CBDDLPLayerDef.get_size())
+                layer_def = CBDDLPLayerDef.unpack(file.read(CBDDLPLayerDef.get_size()))
+                end_byte_offset_by_layer.append(
+                    layer_def.image_offset + layer_def.image_length
+                )
+
+            return CBDDLPFile(
+                filename=path.name,
+                bed_size_mm=(
+                    round(CBDDLP_header.bed_size_x_mm, 4),
+                    round(CBDDLP_header.bed_size_y_mm, 4),
+                    round(CBDDLP_header.bed_size_z_mm, 4),
+                ),
+                height_mm=CBDDLP_header.height_mm,
+                layer_height_mm=CBDDLP_header.layer_height_mm,
+                layer_count=CBDDLP_header.layer_count,
+                resolution=(CBDDLP_header.resolution_x, CBDDLP_header.resolution_y),
+                print_time_secs=CBDDLP_header.print_time,
+                end_byte_offset_by_layer=end_byte_offset_by_layer,
+                slicer_version=".".join(
+                    [
+                        str(CBDDLP_slicer.version_release),
+                        str(CBDDLP_slicer.version_major),
+                        str(CBDDLP_slicer.version_minor),
+                        str(CBDDLP_slicer.version_patch),
+                    ]
+                ),
+                printer_name=printer_name,
+            )
+
+    @classmethod
+    def read_preview(cls, path: pathlib.Path) -> png.Image:
+        with open(str(path), "rb") as file:
+            CBDDLP_header = CBDDLPHeader.unpack(file.read(CBDDLPHeader.get_size()))
+
+            file.seek(CBDDLP_header.high_res_preview_offset)
+            preview = CBDDLPPreview.unpack(file.read(CBDDLPPreview.get_size()))
+
+            file.seek(preview.image_offset)
+            data = file.read(preview.image_length)
+
+            return _read_image(preview.resolution_x, preview.resolution_y, data)
```

Comment:
these are a bit of copy pasta but I'm cool with it for now.

I have some ideas on how we can refactor this